### ISSUE TITLE
superfluous tf.reshape affacts output

### ISSUE THREAD

**Issue by tonyxty** [tensorflow/tensorflow#19948]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 18.04
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: v1.8.0-0-g93bc2e2072 1.8.0
- **Python version**: Python 3.6.5
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: run the following code snippet

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
The code snippet attached, when run with different values for `reshape_iter`, produces different results.

This should not happen since `reshape`-ing a tensor repeatedly to the same shape is, by definition, idempotent.  Could it be that `reshape` affects the internal state of the random number generator?

### Source code / logs
````python
import numpy as np
import tensorflow as tf
import tflearn


def network(x, reshape_iter):
    _, height, width, c_dim = x.shape
    for i in range(reshape_iter):
        x = tf.reshape(x, (-1, height * width * c_dim))
    logits = tflearn.fully_connected(x, 1)
    x = tflearn.activation(logits, 'sigmoid')
    return x, logits


np.random.seed(123456)
tf.set_random_seed(123456)

disc_input = tflearn.input_data(shape=(None, 128, 128, 3))
disc_output, logits = network(disc_input, reshape_iter=5)

sample = np.random.random((8, 128, 128, 3))
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    disc_output_val, logits_val, W_val = sess.run(
            (disc_output, logits, logits.W), feed_dict={disc_input: sample}
    )

print(disc_output_val.reshape((-1,)))
print(W_val.reshape((-1,)))
````

**Comment by mrry** [comment]
> Could it be that reshape affects the internal state of the random number generator?

If a random-number–generating op such as `tf.random_uniform()` is created without an explicit `seed` argument, then TensorFlow [uses the current number of nodes in the graph](https://github.com/tensorflow/tensorflow/blame/master/tensorflow/python/framework/random_seed.py#L68) as part of the seed. Therefore, adding any op, including a `tf.reshape()` prior to a random-number–generating op may change the random numbers that it generates. To prevent this from happening, you must specify an explicit seed for each random-number–generating op.

I'll close this issue as it is working as intended.

**Comment by tonyxty** [comment]
Thanks for the info!
